### ISSUE TITLE
`core::hint::must_use` returns a temporary

### ISSUE THREAD

**Issue by tgross35** [rust-lang/rust#124493]
Originally noticed in https://github.com/rust-lang/rust/issues/124478#issue-2267570324, @GKFX identified it as a problem in `must_use` in https://github.com/rust-lang/rust/issues/124478#issuecomment-2081492040

I tried this code:

```rust
#![feature(hint_must_use)]

fn foo() -> &'static [usize] {
    &[core::hint::must_use(42)]
}
```

I expected to see this happen:  no error, the value is a constant

Instead, this happened: 

```text
error[E0515]: cannot return reference to temporary value
 --> src/lib.rs:4:5
  |
4 |     &[core::hint::must_use(42)]
  |     ^--------------------------
  |     ||
  |     |temporary value created here
  |     returns a reference to data owned by the current function

For more information about this error, try `rustc --explain E0515`.
error: could not compile `playground` (lib) due to 1 previous error
```

### Meta

rustc 1.79 nightly 2024-04-27

**Comment by GKFX** [comment]
Is this likely to be fixable with `must_use` being a function? As I understand it, `foo` only knows the signature of `must_use`, not its implementation, so cannot know that `must_use` is an identity function and cannot avoid a temporary. It would be ideal if the `#[must_use]` attribute applied to expressions.

A magic `#[rustc_identity_function]` attribute on `must_use` could be used to let the compiler know in `foo` that it can avoid the temporary, but seems like overkill.

`git grep 'hint.*must_use'` doesn't show any other usage (besides offset_of) of `core::hint::must_use` which makes me wonder how useful it is. Tagging #94745.

**Comment by tgross35** [comment]
Yup you are right, I guess we can't actually do anything with the function signature so I will close this. Agreed that it may not be overly useful in its current form.

**Comment by tgross35** [comment]
Per  @RalfJung in https://github.com/rust-lang/rust/pull/124484#issuecomment-2092933552, this may indeed be fixable by using `#[rustc_promotable]`

**Comment by dlight** [comment]
Maybe turning `core::hint::must_use` into a macro could fix this issue?
